ssh connection: has_tty respects become plugin require_tty

maylikenoother opened the pull request:
## Summary
Fixes ansible/ansible#86298.

Since 2.19, `ssh` connection [has_tty](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/lib/ansible/plugins/connection/ssh.py:1584:4-1586:39) uses [_is_tty_requested()](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/lib/ansible/plugins/connection/ssh.py:1588:4-1615:20) which only inspects ssh args and ignores the `use_tty` option. This causes become plugins that require a TTY (e.g. `community.general.machinectl`) to fail even when `ansible_ssh_use_tty`/`use_tty` is enabled.

## Changes
- Make [_is_tty_requested()](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/lib/ansible/plugins/connection/ssh.py:1588:4-1615:20) return True when `use_tty` is enabled, so [has_tty](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/lib/ansible/plugins/connection/ssh.py:1584:4-1586:39) matches [exec_command()](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/lib/ansible/plugins/connection/ssh.py:1489:4-1528:43) behavior.
- Add a unit test to ensure [has_tty](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/lib/ansible/plugins/connection/ssh.py:1584:4-1586:39) respects `use_tty`.
- Add changelog fragment.

## Testing
Unit test added; CI will validate.

maylikenoother committed:
ssh connection: has_tty respects use_tty option

s-hertel commented on test/units/plugins/connection/test_ssh.py:
```diff
@@ -70,6 +70,19 @@ def test_plugins_connection_ssh_exec_command(self):
+
+        conn.get_option = MagicMock(side_effect=get_option)
+
+        self.assertTrue(conn.has_tty)
```
An integration test would be better to test the interaction between the task executor, become, and connection plugins.

maylikenoother replied:
Agreed it would be a better end-to-end check of executor+become+connection interaction. If you can point me to a preferred existing integration target to extend (or a recommended pattern), I can follow up with an integration test.

s-hertel commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -1588,6 +1588,9 @@ def has_tty(self):
     def _is_tty_requested(self):
 
+        if self.get_option('use_tty'):
+            return True
```
This breaks pipelining when use_tty is true (which is the default value). I guess there is no test for it, since CI is not failing. How about:

```suggestion
        if self.become and self.become.require_tty:
            return True
```

maylikenoother committed:
Fix ssh has_tty without breaking pipelining

maylikenoother commented:
Thanks for the review!
 
I updated the ssh connection logic to avoid disabling pipelining when `use_tty` is true (default). [_is_tty_requested()](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/lib/ansible/plugins/connection/ssh.py:1588:4-1615:20) now returns True only when the active become plugin requires a TTY (`become.require_tty`) or when explicit SSH args request a TTY (`-t` / `RequestTTY=yes|force`).
 
I also updated/added unit tests:
- [has_tty](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/lib/ansible/plugins/connection/ssh.py:1584:4-1586:39) respects `become.require_tty`
- regression: pipelining is not disabled solely due to default `use_tty`

s-hertel commented on test/units/plugins/connection/test_ssh.py:
```diff
@@ -70,6 +70,29 @@ def test_plugins_connection_ssh_exec_command(self):
+        conn.get_option = MagicMock(side_effect=get_option)
+
+        with patch('ansible.plugins.connection.ConnectionBase.is_pipelining_enabled', return_value=True):
+            self.assertTrue(conn.is_pipelining_enabled())
```
I think the unit tests should be removed in favor of integration tests.

It might be better to check the action uses pipelining. You could add an integration test to https://github.com/ansible/ansible/blob/devel/test/integration/targets/connection_ssh and grep for this [-vvv message](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/__init__.py#L1148) maybe.

```yaml
# test/integration/targets/connection_ssh/test_pipelining.yml
- hosts: ssh-pipelining
  gather_facts: no
  tasks:
    - command: whoami
```

Add something like this to the runme.sh file:
```bash
ansible-playbook test_pipelining.yml -i test_connection.inventory -vvv | grep "Pipelining is enabled."
```

You can run the integration test locally, for example `ansible-test integration connection_ssh --docker ubuntu2404` (see https://docs.ansible.com/projects/ansible/latest/dev_guide/testing_integration.html#quick-start).

bcoca reviewed (changes requested):
I would stress integration tests over unit tests in this case.

s-hertel commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -1588,6 +1588,9 @@ def has_tty(self):
     def _is_tty_requested(self):
 
+        if self.become and getattr(self.become, 'require_tty', False):
+            return True
```
~After thinking about this some more, I think I'm in favor of deleting the `has_tty` property instead of adding this here.~

Sorry for the misguided suggestions. As @bcoca says, checking `use_tty` in `_is_tty_requested` is the right approach, it was just incomplete.

mattclay commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -1588,6 +1588,9 @@ def has_tty(self):
 
     def _is_tty_requested(self):
 
+        if self.become and getattr(self.become, 'require_tty', False):
```
The `require_tty` attribute is on the base become plugin, so use `self.become.require_tty` instead of `getattr`.

maylikenoother replied:
replaced `getattr(self.become, 'require_tty', False)` with `if self.become and self.become.require_tty`

maylikenoother committed:
connection/ssh: cover pipelining via integration test

bcoca commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -1588,6 +1588,9 @@ def has_tty(self):
 
     def _is_tty_requested(self):
 
+        if self.become and self.become.require_tty:
```
As i mention here https://github.com/ansible/ansible/pull/86367#issuecomment-3712353649, the approach itself is incorrect and just hides the underlying problem. There was no need to change how you access the property, using it here at all is incorrect.

mattclay replied:
Sorry, I overlooked that when commenting on the attribute access.

maylikenoother replied:
Thanks – understood. I agree the change I made was treating the symptom.
 
I’ll drop the `become.require_tty` check from [_is_tty_requested()](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/lib/ansible/plugins/connection/ssh.py:1588:4-1615:20) and instead align tty detection with the actual ssh command generation.

maylikenoother committed:
connection/ssh: align has_tty with actual tty allocation

webknjaz commented on test/integration/targets/connection_ssh/runme.sh:
```diff
@@ -75,6 +75,9 @@ ANSIBLE_SSH_TRANSFER_METHOD=piped ./posix.sh "$@"
 ansible-playbook check_ssh_defaults.yml "$@" -i test_connection.inventory
 
+# Ensure pipelining remains enabled by default.
+ansible-playbook test_pipelining.yml "$@" -i test_connection.inventory -vvv | grep 'Pipelining is enabled.'
```
How is this expected to work? Ain't no such playbook in the repo.

maylikenoother replied:
Good catch — there is no `test_pipelining.yml` under [test/integration/targets/connection_ssh/](cci:7://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/test/integration/targets/connection_ssh:0:0-0:0).

I updated [runme.sh](cci:7://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/test/integration/targets/connection_ssh/runme.sh:0:0-0:0) to instead run `ansible -m ping ssh-pipelining -vvvv ... | grep 'Pipelining is enabled.'`, which is where that message actually comes from (action plugin verbosity) and uses the existing `ssh-pipelining` host in [test_connection.inventory](cci:7://file:///c:/Users/brass/OneDrive/Desktop/Work/App/mary-forks/ansible/test/integration/targets/connection_ssh/test_connection.inventory:0:0-0:0) (ansible_ssh_pipelining=true).

maylikenoother committed:
tests: fix connection_ssh pipelining check

maylikenoother committed:
Fix read method calls for stdout and stderr streams

maylikenoother commented:
@webknjaz @bcoca @mattclay @s-hertel 
Please, your reviews are highly appreciated. If you are free, can you drop a review?

s-hertel commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -1245,7 +1245,7 @@ def _bare_run(self, cmd: list[bytes], in_data: bytes | None, sudoable: bool = Tr
                 for key, event in events:
                     if key.fileobj == p.stdout:
-                        b_chunk = p.stdout.read()
+                        b_chunk = p.stdout.read() if hasattr(p.stdout, 'read') else b''
```
What is this change for?

s-hertel commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -1260,7 +1260,7 @@ def _bare_run(self, cmd: list[bytes], in_data: bytes | None, sudoable: bool = Tr
                         display.debug(u"stdout chunk (state=%s):\n>>>%s<<<\n" % (state, to_text(b_chunk)))
                     elif key.fileobj == p.stderr:
-                        b_chunk = p.stderr.read()
+                        b_chunk = p.stderr.read() if hasattr(p.stderr, 'read') else b''
```
Same question here.

s-hertel commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -1511,10 +1511,8 @@ def exec_command(self, cmd: str, in_data: bytes | None = None, sudoable: bool =
-
         args: tuple[str, ...]
-        if not in_data and sudoable and use_tty:
+        if self._is_tty_requested(in_data=in_data is not None, sudoable=sudoable):
```
Checking if `in_data is not None` instead of `not in_data` is a change in behavior.
```suggestion
        if self._is_tty_requested(in_data=bool(in_data), sudoable=sudoable):
```

s-hertel commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -1586,13 +1584,13 @@ def close(self) -> None:
         for opt in ('ssh_args', 'ssh_common_args', 'ssh_extra_args'):
             attr = self.get_option(opt)
-            if attr is not None:
+            if isinstance(attr, str):
```
What's this change for?

s-hertel commented on test/units/plugins/connection/test_ssh.py:
```diff
@@ -70,6 +70,15 @@ def test_plugins_connection_ssh_exec_command(self):
+        conn.become = MagicMock()
+        conn.become.require_tty = True
+
+        self.assertTrue(conn.has_tty)
```
This will not prevent a regression, so I think the unit tests should be removed in favor of integration tests.

You can run the integration test locally, for example ansible-test integration connection_ssh --docker ubuntu2404 (see https://docs.ansible.com/projects/ansible/latest/dev_guide/testing_integration.html#quick-start).

s-hertel commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -1610,12 +1608,17 @@ def _is_tty_requested(self):
         """ override parent method and ensure we don't request a tty """
 
-        if self._is_tty_requested():
+        if self._is_tty_requested(in_data=True, sudoable=True):
```
I have a feeling that this is not correct. Also, there's no reason to pass `sudoable=True` since that is the default.

s-hertel reviewed (commented):
I haven't given this an in-depth review, but had some questions. Take my review with a grain of salt (except the request for better tests, please).

If a become plugin sets `require_tty=True`, I think it should automatically disable pipelining.

maylikenoother committed:
ssh connection: disable pipelining when become requires TTY

maylikenoother committed:
ssh connection: revert test-only defensive guards

maylikenoother committed:
tests: avoid unnecessary lambda in ssh connection unit test

maylikenoother committed:
tests: assert ssh tty selection honors use_tty
